lib/core/src/Cardano/Wallet/DB/SqliteTypes.hs

```diff
@@ -0,0 +1,246 @@
+import Web.PathPieces
+
+import qualified Data.Text as T
+
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..)
+    , Direction (..)
+    , Hash (..)
+    , SlotId (..)
+    , TxStatus (..)
+    , WalletId (..)
+    , WalletState (..)
+    , flatSlot
+    , fromFlatSlot
+    , isValidCoin
+    )
+
+----------------------------------------------------------------------------
+-- Helper functions
+
+parseUrlPieceFromText :: FromText a => Text -> Either Text a
+parseUrlPieceFromText = first (T.pack . getTextDecodingError) . fromText
+
+aesonFromText :: FromText a => String -> Value -> Parser a
+aesonFromText what = withText what $ either (fail . show) pure . fromText
+
+----------------------------------------------------------------------------
+-- Direction
+
+instance PersistField Direction where
+    toPersistValue = toPersistValue . directionToBool
+    fromPersistValue pv = do
+        let err = T.pack $ "not a valid value: " <> show pv
+        bimap (const err) directionFromBool (fromPersistValue pv)
+
+instance PersistFieldSql Direction where
+    sqlType _ = sqlType (Proxy @Bool)
+
+directionToBool :: Direction -> Bool
+directionToBool Incoming = True
+directionToBool Outgoing = False
+
+directionFromBool :: Bool -> Direction
+directionFromBool True = Incoming
+directionFromBool False = Outgoing
+
+----------------------------------------------------------------------------
+-- WalletId
+
+instance PersistField WalletId where
+    toPersistValue =
+        toPersistValue . toText
+    fromPersistValue pv = do
+        a <- fromText <$> fromPersistValue pv
+        case a of
+             Left _     ->
+                 Left . T.pack $ "not a valid value: " <> show pv
+             Right txid ->
+                 pure txid
+
+instance PersistFieldSql WalletId where
+    sqlType _ = sqlType (Proxy @Text)
+
+instance Read WalletId where
+    readsPrec _ = error "readsPrec stub needed for persistent"
+
+instance ToHttpApiData WalletId where
+    toUrlPiece = toText
+
+instance FromHttpApiData WalletId where
+    parseUrlPiece = parseUrlPieceFromText
+
+instance ToJSON WalletId where
+    toJSON = String . toText
+
+instance FromJSON WalletId where
+    parseJSON = aesonFromText "WalletId"
+
+instance PathPiece WalletId where
+    fromPathPiece = fromTextMaybe
+    toPathPiece = toText
+
+
+----------------------------------------------------------------------------
+-- AddressScheme
+
+data AddressScheme = Sequential | Random deriving (Show, Eq, Generic)
+
+instance PersistField AddressScheme where
+    toPersistValue = toPersistValue . schemeToBool
+    fromPersistValue pv = do
+        let err = T.pack $ "not a valid value: " <> show pv
+        bimap (const err) schemeFromBool (fromPersistValue pv)
+
+instance PersistFieldSql AddressScheme where
+    sqlType _ = sqlType (Proxy @Bool)
+
+schemeToBool :: AddressScheme -> Bool
+schemeToBool Sequential = True
+schemeToBool Random = False
+
+schemeFromBool :: Bool -> AddressScheme
```

Comment:
Perhaps this deserves a comment, to explain the exact mapping that will take place. (Important for someone reading the haddock documentation.)

Reply:
@paweljakubas

Reply:
will push refined AddressScheme in a second

Reply:
Just to clarify. If I'm a casual user of this function, it currently isn't clear how a given `Bool` value will map onto an `AddressScheme`. There are two possible mappings:

1. `{True, False}` -> `{Sequential, Random}
2. `{False, True}` -> `{Sequential, Random}`

To understand precisely which mapping is being used, I'd have to dig into the source code, which kind of defeats the point.

I would recommend either:

1. Removing this function (as this mapping is somewhat arbitrary).
2. Renaming it (so that the exact mapping performed is clear from the name).
3. Adding a haddock documentation comment to make the mapping explicit.

Reply:
Introduced `data AddressScheme = Sequential | Random | Any` now relying on `FromText`, `ToText` - so basically opted for point 1 and 2 ;-)